### PR TITLE
replicate delete branch

### PR DESCRIPTION
**Pull request by max-hoffman**
Replicate will push deletes on write in syncronous and async mode. Read replicas will delete refs that don't exist on the replica. Both of these can have destructive effects if not used carefully.

### CODE REVIEW

File: integration-tests/bats/replication.bats [hunk 1]
```diff
@@ -58,6 +58,32 @@ teardown() {
...
+    [ "$status" -eq 0 ]
+    [ "${#lines[@]}" -eq 2 ]
+    [[ "$output" =~ "t1" ]] || false
+
+    cd ../repo1
+    dolt checkout main
+    dolt sql -q "call dolt_branch('-df', 'feature');"
+
+    cd ..
+    dolt clone file://./bac1 repo3
```
Comment (timsehn) [inline integration-tests/bats/replication.bats:82]: I think you may need an additional test, potentially skipped, where the branch already exists on the replica.

Comment (max-hoffman) [reply integration-tests/bats/replication.bats]: added more tests

File: go/libraries/doltcore/sqle/read_replica_database.go [hunk 2]
```diff
@@ -230,6 +223,58 @@ func pullBranches(ctx *sql.Context, rrd ReadReplicaDatabase, branches []string)
 	return nil
 }
 
+func getReplicationBranches(ctx *sql.Context, rrd ReadReplicaDatabase) ([]string, []ref.DoltRef, error) {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/read_replica_database.go:226]: Should generally name return params when they are ambiguous like this

File: go/libraries/doltcore/sqle/read_replica_database.go [hunk 3]
```diff
@@ -230,6 +223,58 @@ func pullBranches(ctx *sql.Context, rrd ReadReplicaDatabase, branches []string)
...
+			deletedBranches = append(deletedBranches, localRefs[j])
+			j++
+		}
+	}
+	for i < len(srcRefs) {
+		allBranches = append(allBranches, srcRefs[i].GetPath())
+		i++
+	}
+	for j < len(localRefs) {
+		deletedBranches = append(deletedBranches, localRefs[i])
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/read_replica_database.go:260]: `i` is never being incremented here, this is suspicious

This algorithm could use a quick unit test (deleted branches at the beginning, at the end, in the middle)

**Review by zachmu — APPROVED** [review]
LGTM

File: go/libraries/doltcore/sqle/read_replica_database.go (hunk 3 above)
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/read_replica_database.go:260]: yeah that aint right, thanks

**Pull request merged**